chain/chain/src/chain.rs

```diff
@@ -4755,15 +4759,22 @@ impl<'a> ChainUpdate<'a> {
         // the delta for the shards that we are tracking this epoch
         #[cfg(feature = "protocol_feature_flat_state")]
         if self.runtime_adapter.cares_about_shard(me.as_ref(), &prev_hash, shard_id, true) {
+            let delta = FlatStateDelta::from_state_changes(&trie_changes.state_changes());
+
             if let Some(chain_flat_storage) =
                 self.runtime_adapter.get_flat_storage_state_for_shard(shard_id)
             {
-                let delta = FlatStateDelta::from_state_changes(&trie_changes.state_changes());
                 let block_info = flat_state::BlockInfo { hash: block_hash, height, prev_hash };
                 let store_update = chain_flat_storage
                     .add_block(&block_hash, delta, block_info)
                     .map_err(|e| StorageError::from(e))?;
                 self.chain_store_update.merge(store_update);
+            } else {
+                info!(target: "chain", %shard_id, "Add delta for flat storage creation");
+                let mut store_update = self.chain_store_update.store().store_update();
+                store_helper::set_delta(&mut store_update, shard_id, block_hash.clone(), &delta)
+                    .map_err(|e| StorageError::from(e))?;
+                self.chain_store_update.merge(store_update);
```

Comment:
I was thinking that we can do gc during FS head catchup, but then we would miss deltas outside of canonical chain.
Then I agree that adding it to regular gc is better, it will be done in following PRs.